lib/cdo/throttle.rb

```diff
@@ -0,0 +1,45 @@
+require 'cdo/shared_cache'
+require 'dynamic_config/dcdo'
+
+module Cdo
+  module Throttle
+    CACHE_PREFIX = "cdo_throttle/".freeze
+
+    # @param [String] id - Unique identifier to throttle on.
+    # @param [Integer] limit - Number of requests allowed over period.
+    # @param [Integer] period - Period of time in seconds.
+    # @param [Integer] throttle_for - How long id should stay throttled in seconds. Optional.
+    # Defaults to Cdo::Throttle.throttle_time.
+    # @returns [Boolean] Whether or not the request should be throttled.
+    def self.throttle(id, limit, period, throttle_for = throttle_time)
+      full_key = CACHE_PREFIX + id.to_s
+      value = CDO.shared_cache.read(full_key) || empty_value
+      now = Time.now.utc
+      value[:request_timestamps] << now
+
+      if value[:throttled_until]&.future?
+        should_throttle = true
+      else
+        value[:throttled_until] = nil
+        earliest = now - period
+        value[:request_timestamps].select! {|timestamp| timestamp >= earliest}
+        should_throttle = value[:request_timestamps].size > limit
+        value[:throttled_until] = now + throttle_for if should_throttle
+      end
+
+      CDO.shared_cache.write(full_key, value)
```

Comment:
There's a race condition here if several different processes are all reading from and writing to the value at the same time which will undercount the number of requests.  (That might be acceptable given the use case, but it would be a good limitation to note.)

Reply:
ah, yeah, you're right. i think it's acceptable for the current consumers (the client is caching responses and the server is doing some caching as well, so this is only called for unique/uncached requests), but i'd like to fix this to future-proof it.

i think the solution for this would be to implement a [mutex](https://ruby-doc.org/core-2.5.0/Mutex.html) or a queue here, but are there other/better solutions? i worry about a mutex because i think it would have to lock for everybody (rather than being able to lock per ID), but maybe that's okay -- i'm not super familiar with implementing/using locks in this way

Reply:
If we're using Redis, then this could be a good way to implement it: https://redislabs.com/redis-best-practices/basic-rate-limiting/.  (The behavior would be a bit different as you've defined it here though.)

A local lock probably wouldn't help because it's still local to an instance and a distributed lock is difficult to implement.  Two viable patterns are atomic operations (such as increment) at the cache server or optimistic concurrency (where you make sure the value that you're writing hasn't changed since you read it).